### ISSUE TITLE
WIP note: RMSNorm remaining fixes (local branch rmsnorm-continue)

### ISSUE THREAD

**Issue by bhaswata08** [stanford-cs336/assignment1-basics#93]
## Continue on home PC - RMSNorm remaining work

**Branch:** `rmsnorm-continue` on `bhaswata08/assignment1-basics` (commit 7e71fa8)
**File:** `cs336_basics/transformers/layers.py:81-105`

### Current state
- `Linear` at `cs336_basics/transformers/layers.py:12` and `Embedding` at `cs336_basics/transformers/layers.py:43` done and wired in `tests/adapters.py`
- `RMSNorm` has `gamma` as `nn.Parameter(torch.ones(d_model))` at `cs336_basics/transformers/layers.py:95` with correct `dtype`/`device` handling, and upcast/downcast at `cs336_basics/transformers/layers.py:101-105`

### What is still broken / TODO

- [ ] **Fix denominator - use RMS not std** at `cs336_basics/transformers/layers.py:103-104`
  - Spec is `cs336_assignment1_basics.md:737`: `RMS(a) = sqrt(1/d_model * sum(a_i^2) + eps)`
  - Currently uses `reduce(..., reduction=torch.std)` then `sqrt(std + eps)`. `torch.std` subtracts mean, RMS must not. Replace with mean of squares.
  - `eps` must be inside `sqrt(mean_sq + eps)` not `sqrt(std)+eps`
  - Hint in handout: upcast to float32 before squaring to avoid overflow - already done, keep it

- [ ] **Param name should be `weight` not `gamma`** at `cs336_basics/transformers/layers.py:95`
  - Checkpoint key is `layers.1.ln1.weight` per `tests/test_model.py:178`
  - `state_dict()` currently exposes `gamma` -> `load_state_dict({"weight": ...})` will fail and serialization for Transformer block will mismatch
  - Check with `list(RMSNorm(d_model=4).state_dict().keys())` should be `["weight"]`
  - Store `d_model`/`eps` as `self.d_model`, `self.eps` if needed

- [ ] **Support arbitrary leading dims `... d_model`** 
  - Spec and adapter `tests/adapters.py:384` allow `Float[Tensor, "... d_model"]`, not just `batch seq d_model` at `cs336_basics/transformers/layers.py:100`
  - Current `reduce` pattern `"batch_size seq_len d_model -> batch_size seq_len"` and `rearrange("b s -> b s 1")` at `cs336_basics/transformers/layers.py:103-104` hardcodes 2 leading dims
  - Should normalize over last dim only and broadcast `weight` across any leading dims

- [ ] **Broadcast `weight` correctly** at `cs336_basics/transformers/layers.py:104`
  - Equation 4: `a_i / RMS(a) * g_i` - elementwise multiply after norm. Verify `x * weight` broadcasts when `x` is `... d_model` and `weight` is `d_model`

- [ ] **Wire `tests/adapters.py:371` `run_rmsnorm`**
  - Should construct `RMSNorm(d_model, eps, device=..., dtype=...)`, load `weights` via `load_state_dict` or `weight.data`, return forward output
  - Reference: how `run_linear`/`run_embedding` are wired at `tests/adapters.py:31` and `tests/adapters.py:60`

- [ ] **Sanity checks before pushing further**
  - Toy: `d_model=3`, `x=[1,2,3]`, `g=1` compute expected `x / sqrt(mean(x^2)+eps)` by hand vs module output
  - Edge: `x=[2,2,2]` -> `RMS=2`, output should be `[1,1,1]` with `g=1`. `torch.std` gives 0 here - good canary for bug
  - `uv run pytest -k test_rmsnorm` should pass `tests/test_model.py:176` against snapshot
  - Check `pytest -k "test_linear or test_embedding"` still pass after rename

### How to verify on home PC
```bash
git fetch fork
git checkout rmsnorm-continue
uv run pytest -k test_rmsnorm -v
uv run pytest -k "test_linear or test_embedding" -v
```

### References
- Handout RMSNorm spec `cs336_assignment1_basics.md:733-771`
- Adapter contract `tests/adapters.py:371-391`
- Test `tests/test_model.py:176-183`

**Comment by bhaswata08** [comment]
Opened on upstream by mistake - tracking moved to fork bhaswata08/assignment1-basics#1
